### PR TITLE
Remove uses of `typeof` and the `?:` operator

### LINKED ISSUE #846: Remove all usage of typeof() and change .xcconfig to disallow GCC extensions
typeof() is a GCC extension that we don't want to support in the code base.

Internal build settings disallow it by specifying that the compiler should use C99, not C99+GCC extensions, but it seems that MDCCatalog does not do so.

### PR DESCRIPTION
**Pull request by romoore**
Make the code base more portable by replacing some GNU extension code
with C99-compatible versions.

* Replace shortcut `?:` operator with explicit middle (*true*) terms
* Remove pointer inequality checks in conditionals
* (Cleanup) Change to `#include` for `<tgmath.h>` based on style guide

Closes #846

### CODE REVIEW

File: components/FlexibleHeader/src/MDCFlexibleHeaderView.m [hunk 1]
```diff
@@ -662,24 +663,29 @@ - (void)fhv_accumulatorDidChange {
...
       // When in front of infinite content we only care to hide the shadow when our header is
       // off-screen.
-      shadowIntensity = MAX(0, MIN(1, accumulator / kShadowScaleLength));
+      shadowIntensity = MDCFMax(0, MDCFMin(1, accumulator / kShadowScaleLength));
 
     } else {
       // When over non-infinite content we also want to hide the shadow when we're anchored to the
       // top of our content.
-      shadowIntensity = MAX(0, MIN(1, MIN(accumulator, frameBottomEdge) / kShadowScaleLength));
+      shadowIntensity = MDCFMax(0,
```
Comment (ianegordon) [inline components/FlexibleHeader/src/MDCFlexibleHeaderView.m]: If you add a linebreak after "shadowIntensity =" will everything fit on the next line?

File: components/PageControl/src/MDCPageControl.m [hunk 2]
```diff
@@ -56,6 +58,16 @@ static inline CGFloat normalizeValue(CGFloat value, CGFloat minRange, CGFloat ma
   return (diff > 0) ? ((value - minRange) / diff) : 0;
 }
 
+/**
+ Snap @c value to the range <code>[min, max]</code>.
+ */
+static inline NSInteger snapToRange(NSInteger min, NSInteger max, NSInteger value) {
```
Comment (ianegordon) [inline components/PageControl/src/MDCPageControl.m]: Functions are typically capitalized.
For minimal surprise, consider naming Clamp(value, min, max) to match other implementations.

File: components/TextFields/src/MDCTextInputControllerFullWidth.m [hunk 3]
```diff
@@ -849,7 +851,7 @@ - (void)setErrorText:(NSString *)errorText
 
     NSString *valueString = @"";
 
-    if (self.textInput.text > 0) {
+    if (self.textInput.text) {
```
Comment (ianegordon) [inline components/TextFields/src/MDCTextInputControllerFullWidth.m]: Was this originally intended to be:
if (self.textInput.text.length > 0)

**Review by ianegordon — CHANGES_REQUESTED** [review]
MIN and MAX are defined in NSObjCRuntime.h which is in Foundation.
I'm not sure we should be re-implementing it.
Let's discuss offline.

File: components/FlexibleHeader/src/MDCFlexibleHeaderView.m (hunk 1 above)
Comment (romoore) [reply components/FlexibleHeader/src/MDCFlexibleHeaderView.m]: Yep, will change.

File: components/PageControl/src/MDCPageControl.m (hunk 2 above)
Comment (romoore) [reply components/PageControl/src/MDCPageControl.m]: Done.

**Comment by ianegordon** [comment]
Discussed offline.
TODO: Investigate if it is simple to enable the extension warning on our code without tripping up on the system code.
Move back to MIN / MAX.

File: components/TextFields/src/MDCTextField.m [hunk 4]
```diff
@@ -56,8 +56,9 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
     NSString *interfaceBuilderPlaceholder = super.placeholder;
     [self commonMDCTextFieldInitialization];
 
-    _fundament = [aDecoder decodeObjectForKey:MDCTextFieldFundamentKey]
-                     ?: [[MDCTextInputCommonFundament alloc] initWithTextInput:self];
+    MDCTextInputCommonFundament *fundament = [aDecoder decodeObjectForKey:MDCTextFieldFundamentKey];
+    _fundament = fundament
+                     ? fundament : [[MDCTextInputCommonFundament alloc] initWithTextInput:self];
```
Comment (ianegordon) [inline components/TextFields/src/MDCTextField.m]: Consider a linebreak after the "lvar =" to keep the ternary expression on a single line.

File: components/TextFields/src/MDCTextField.m [hunk 5]
```diff
@@ -378,7 +379,7 @@ - (CGSize)intrinsicContentSize {
 
   CGFloat underlineLabelsHeight =
       MAX(MDCCeil(self.leadingUnderlineLabel.font.lineHeight * 2.f) / 2.f,
-          MDCCeil(self.trailingUnderlineLabel.font.lineHeight * 2.f) / 2.f);
+              MDCCeil(self.trailingUnderlineLabel.font.lineHeight * 2.f) / 2.f);
```
Comment (ianegordon) [inline components/TextFields/src/MDCTextField.m]: Not sure about this indenting.  +4 spaces after continuing a line.  This looks like +8.

File: components/ButtonBar/src/private/MDCAppBarButtonBarBuilder.m [hunk 6]
```diff
@@ -82,7 +82,7 @@ - (UIView *)buttonBar:(MDCButtonBar *)buttonBar
   [self transferCustomViewOwnershipForBarButtonItem:buttonItem];
 
   // Take the real custom view if it exists instead of sandbag view.
-  UIView *customView = buttonItem.mdc_customView ?: buttonItem.customView;
+  UIView *customView = buttonItem.mdc_customView ? buttonItem.mdc_customView : buttonItem.customView;
```
Comment (ianegordon) [inline components/ButtonBar/src/private/MDCAppBarButtonBarBuilder.m]: 101 characters :/
Linebreak after "customView ="

File: components/Slider/src/MDCSlider.m [hunk 7]
```diff
@@ -96,7 +96,7 @@ - (void)commonMDCSliderInit {
 #pragma mark - ThumbTrack passthrough methods
 
 - (void)setTrackBackgroundColor:(UIColor *)trackBackgroundColor {
-  _thumbTrack.trackOffColor = trackBackgroundColor ?: [[self class] defaultTrackOffColor];
+  _thumbTrack.trackOffColor = trackBackgroundColor ? trackBackgroundColor : [[self class] defaultTrackOffColor];
```
Comment (ianegordon) [inline components/Slider/src/MDCSlider.m]: Linebreak after ".trackOffColor =".

File: components/TextFields/src/MDCTextInputControllerDefault.m [hunk 8]
```diff
@@ -493,7 +493,8 @@ - (void)movePlaceholderToUp:(BOOL)isToUp {
       self.textInput.placeholderLabel.transform = CGAffineTransformIdentity;
 
       self.textInput.placeholderLabel.textColor =
-          self.previousPlaceholderColor ?: self.inlinePlaceholderColor;
+          self.previousPlaceholderColor
+            ? self.previousPlaceholderColor : self.inlinePlaceholderColor;
```
Comment (ianegordon) [inline components/TextFields/src/MDCTextInputControllerDefault.m]: I'm beginning to question the utility of the ternary operator without an implicit second argument.  Once you end up on multiple lines it gets hard to parse.

Consider:
if (self.previousPlaceholderColor) {
  self.textInput.placeholderLabel.textColor = self.previousPlaceholderColor;
} else {
  self.textInput.placeholderLabel.textColor = self.inlinePlaceholderColor;
}

File: components/TextFields/src/MDCTextInputControllerDefault.m [hunk 9]
```diff
@@ -1044,7 +1047,7 @@ - (void)setErrorText:(NSString *)errorText
 
     NSString *valueString = @"";
 
-    if (self.textInput.text > 0) {
+    if (self.textInput.text) {
```
Comment (ianegordon) [inline components/TextFields/src/MDCTextInputControllerDefault.m]: Was the original intention (self.textInput.text.length > 0) ?

**Review by ianegordon — CHANGES_REQUESTED** [review]
I find ternary operators split across lines very hard to parse.  Can we discuss this offline?

File: components/TextFields/src/MDCTextField.m (hunk 4 above)
Comment (romoore) [reply components/TextFields/src/MDCTextField.m]: Done

File: components/TextFields/src/MDCTextField.m (hunk 5 above)
Comment (romoore) [reply components/TextFields/src/MDCTextField.m]: Done

File: components/TextFields/src/MDCTextInputControllerDefault.m (hunk 8 above)
Comment (romoore) [reply components/TextFields/src/MDCTextInputControllerDefault.m]: Done

File: components/Slider/src/MDCSlider.m (hunk 7 above)
Comment (romoore) [reply components/Slider/src/MDCSlider.m]: Done

File: components/ButtonBar/src/private/MDCAppBarButtonBarBuilder.m (hunk 6 above)
Comment (romoore) [reply components/ButtonBar/src/private/MDCAppBarButtonBarBuilder.m]: Done

File: components/TextFields/src/MDCTextInputControllerDefault.m (hunk 9 above)
Comment (willlarche) [reply components/TextFields/src/MDCTextInputControllerDefault.m]: Typo. Should be .length

File: components/TextFields/src/MDCTextInputControllerFullWidth.m (hunk 3 above)
Comment (willlarche) [reply components/TextFields/src/MDCTextInputControllerFullWidth.m]: yesy

**Review by willlarche — APPROVED** [review]

**Review by ianegordon — APPROVED** [review]
